### PR TITLE
Avoid raw pointers in projecting implementation

### LINKED ISSUE #84: Look into attached unit test logs
While #83, this popped up (summary, complete logs attached)

```
5: Indirect leak of 40 byte(s) in 1 object(s) allocated from:                                                                                     
5:     #0 0x7a2c008e1359 in __interceptor_malloc /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69                                
5:     #1 0x5c5933bf1fdc in Eigen::internal::handmade_aligned_malloc(unsigned long, unsigned long) /usr/include/eigen3/Eigen/src/Core/util/Memory.
h:105                                                                                                                                             
5:     #2 0x5c5933bf244b in Eigen::internal::aligned_malloc(unsigned long) /usr/include/eigen3/Eigen/src/Core/util/Memory.h:188                   
5:     #3 0x5c5933d85d8e in void* Eigen::internal::conditional_aligned_malloc<true>(unsigned long) /usr/include/eigen3/Eigen/src/Core/util/Memory.
h:241                                                                                                                                             
5:     #4 0x5c5933d1e568 in double* Eigen::internal::conditional_aligned_new_auto<double, true>(unsigned long) /usr/include/eigen3/Eigen/src/Core/
util/Memory.h:404                                                                                                                                 5:     #5 0x5c5933cb5949 in Eigen::DenseStorage<double, -1, -1, 1, 0>::DenseStorage(Eigen::DenseStorage<double, -1, -1, 1, 0> const&) /usr/include
/eigen3/Eigen/src/Core/DenseStorage.h:589                                                                                                         5:     #6 0x5c5933c7ce30 in Eigen::PlainObjectBase<Eigen::Matrix<double, -1, 1, 0, -1, 1> >::PlainObjectBase(Eigen::PlainObjectBase<Eigen::Matrix<
double, -1, 1, 0, -1, 1> > const&) /usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:512                                                       5:     #7 0x5c5933c46159 in Eigen::Matrix<double, -1, 1, 0, -1, 1>::Matrix(Eigen::Matrix<double, -1, 1, 0, -1, 1> const&) /usr/include/eigen3/Eige
n/src/Core/Matrix.h:414                                                                                                                           5:     #8 0x5c5933c002da in tapkee::MatrixProjectionImplementation::MatrixProjectionImplementation(Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen
::Matrix<double, -1, 1, 0, -1, 1>) tapkee/include/tapkee/projection.hpp:46                                                    5:     #9 0x5c5933d0cf69 in tapkee::tapkee_internal::ImplementationBase<__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > >, tapkee::eigen_kernel_callback, tapkee::eigen_distance_callback, tapkee::eigen_features_callback>::embedPCA() tapkee/include
/tapkee/methods.hpp:341                                                                                                                           5:     #10 0x5c5933caa55c in tapkee::tapkee_internal::ImplementationBase<__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > >, tapkee::eigen_kernel_callback, tapkee::eigen_distance_callback, tapkee::eigen_features_callback>::embedUsing(tapkee::DimensionReductionMethod)
 tapkee/include/tapkee/methods.hpp:113                                                                                        5:     #11 0x5c5933c735ea in tapkee::TapkeeOutput tapkee::embed<__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > >, tapkee::eigen_kernel_callback, tapkee::eigen_distance_callback, tapkee::eigen_features_callback>(__gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > >, __gnu_cxx::__normal_iterator<int*, std::vector<int, std::allocator<int> > >, tapkee::eigen_kernel_callback, tapkee::eigen_d
istance_callback, tapkee::eigen_features_callback, stichwort::ParametersSet) tapkee/include/tapkee/embed.hpp:117              5:     #12 0x5c5933c3f289 in tapkee::tapkee_internal::ParametersInitializedState::embedUsing(Eigen::Matrix<double, -1, -1, 0, -1, -1> const&) cons
t tapkee/tapkee/include/tapkee/chain_interface.hpp:479                                                                               
5:     #13 0x5c5933bdf964 in Projecting_PCA_Test::TestBody() tapkee/tapkee/test/unit/projecting.cpp:17                               
5:     #14 0x7a2c0100465b  (/usr/lib/libgtest.so.1.14.0+0x5365b)                          
5:                                                                                                                                                
5: SUMMARY: AddressSanitizer: 304 byte(s) leaked in 6 allocation(s).                                                                              
5/5 Test #5: projecting .......................***Failed    0.17 sec 
```
[ctest.log](https://github.com/lisitsyn/tapkee/files/14931136/ctest.log)

Hotfix below didn't improve if I didn't overlook results:
```
--- a/test/unit/projecting.cpp
+++ b/test/unit/projecting.cpp
@@@ -13,7 -13,7 +13,8 @@@ TEST(Projecting, PCA
  
      TapkeeOutput output;
  
--    ASSERT_NO_THROW(output = tapkee::initialize().withParameters((method = PCA, target_dimension = 2)).embedUsing(X));
++    const auto tapkee_obj = tapkee::initialize().withParameters((method = PCA, target_dimension = 2));
++    ASSERT_NO_THROW(output = tapkee_obj.embedUsing(X));
  
      auto projected = output.projection(X.col(0));
      ASSERT_EQ(2, projected.size())
```

I didn't get around it yet as I would have liked, so I added it here. From tomorrow I have more time again and I am aiming at getting it going this week. 🤓

### PR DESCRIPTION
**Pull request by lisitsyn**
Probably fixes #84

### CODE REVIEW

**Comment by lisitsyn** [comment]
hey @iglesias would you like to take a look? thanks

**Comment by iglesias** [comment]
Hey! I run and it fixed 2 of the errors. There's still this one (summarized):

```
2: ==191523==ERROR: AddressSanitizer: requested allocation size 0x2d79883d20010 (0x2d79883d21010 after adjustments for alignment, red zones etc.) exceeds maximum supported size of 0x10000000000 (thread T0)
[...]
2:     #7 0x651ab7876242 in void Eigen::PlainObjectBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> >::_init2<int, int>(long, long, Eigen::internal::enable_if<true, int>::type*) /usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:810
2:     #8 0x651ab7837f9f in Eigen::Matrix<double, -1, -1, 0, -1, -1>::Matrix<int, int>(int const&, int const&) /usr/include/eigen3/Eigen/src/Core/Matrix.h:340
2:     #9 0x651ab79b0321 in Eigen::Matrix<double, -1, -1, 0, -1, -1> tapkee::tapkee_internal::compute_distance_matrix<__gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, float_distance_callback>(__gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, __gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, float_distance_callback) tapkee/include/tapkee/routines/multidimensional_scaling.hpp:114
2:     #10 0x651ab7917b97 in tapkee::tapkee_internal::ImplementationBase<__gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, tapkee::dummy_kernel_callback<float>, float_distance_callback, tapkee::dummy_features_callback<float> >::embedMultidimensionalScaling() tapkee/include/tapkee/methods.hpp:195
2:     #11 0x651ab78a95bb in tapkee::tapkee_internal::ImplementationBase<__gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, tapkee::dummy_kernel_callback<float>, float_distance_callback, tapkee::dummy_features_callback<float> >::embedUsing(tapkee::DimensionReductionMethod) tapkee/include/tapkee/methods.hpp:104
2:     #12 0x651ab78643d2 in tapkee::TapkeeOutput tapkee::embed<__gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, tapkee::dummy_kernel_callback<float>, float_distance_callback, tapkee::dummy_features_callback<float> >(__gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, __gnu_cxx::__normal_iterator<float*, std::vector<float, std::allocator<float> > >, tapkee::dummy_kernel_callback<float>, float_distance_callback, tapkee::dummy_features_callback<float>, stichwort::ParametersSet) tapkee/include/tapkee/embed.hpp:117
[...]
2: ==191523==HINT: if you don't care about these errors you may set allocator_may_return_null=1
2: SUMMARY: AddressSanitizer: allocation-size-too-big /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69 in __interceptor_malloc
```
I doubt it is relevant though. It passes without the sanitizer and there's no kind of out of memory error.  And the name of the test fits with it having this type of error ':-D

[tapkee.log](https://github.com/lisitsyn/tapkee/files/14948114/tapkee.log)

File: include/tapkee/projection.hpp [hunk 1]
```diff
@@ -22,12 +24,14 @@ struct ProjectionImplementation
 //! A pimpl wrapper for projecting function
 struct ProjectingFunction
 {
-    ProjectingFunction() : implementation(NULL){};
-    ProjectingFunction(ProjectionImplementation* impl) : implementation(impl){};
-    //! Destroys current implementation
-    void clear()
+    ProjectingFunction() : implementation(NULL)
```
Comment (iglesias) [inline include/tapkee/projection.hpp:27]: A minor one, but let's use `nullptr` instead?

**Comment by iglesias** [comment]
lgtm :ship:

File: include/tapkee/projection.hpp [hunk 2]
```diff
@@ -22,12 +24,14 @@ struct ProjectionImplementation
...
+    {
+    }
+    ProjectingFunction(ProjectionImplementation* impl) : implementation(impl)
+    {
+    }
+    ProjectingFunction(const ProjectingFunction& other) : implementation(other.implementation)
```
[the remark this answers is not in the document]
Comment (lisitsyn) [reply include/tapkee/projection.hpp:33]: Yes, right. Going to remove that.

File: include/tapkee/projection.hpp (hunk 1 above)
Comment (lisitsyn) [reply include/tapkee/projection.hpp:27]: Replacing..

**Review by iglesias — APPROVED** [review]

**Pull request merged**